Mathlib/Algebra/BigOperators/Order.lean

```diff
@@ -867,4 +867,54 @@ def evalFinsetSum : PositivityExt where eval {u α} zα pα e := do
       return .nonnegative q(@sum_nonneg $ι $α $pα' $f $s fun i _ ↦ $pr i)
   | _ => throwError "not Finset.sum"
 
+private alias ⟨_, prod_ne_zero⟩ := prod_ne_zero_iff
+
+/-- The `positivity` extension which proves that `∏ i in s, f i` is nonnegative if `f` is, and
+positive if each `f i` is.
+
+TODO: The following example does not work
+```
+example (s : Finset ℕ) (f : ℕ → ℤ) (hf : ∀ n, 0 ≤ f n) : 0 ≤ s.prod f := by positivity
+```
+because `compareHyp` can't look for assumptions behind binders.
+-/
+@[positivity Finset.prod _ _]
+def evalFinsetProd : PositivityExt where eval {u α} zα pα e := do
+  match e with
+  | ~q(@Finset.prod _ $ι $instα $s $f) =>
+    let i : Q($ι) ← mkFreshExprMVarQ q($ι) .syntheticOpaque
+    have body : Q($α) := Expr.betaRev f #[i]
+    let rbody ← core zα pα body
+    -- Try to show that the sum is positive
+    try
+      let .positive pbody := rbody | failure -- Fail if the body is not provably positive
+      -- TODO: We must name the following, else `assertInstancesCommute` loops.
+      let _instαmon ← synthInstanceQ q(CommMonoidWithZero $α)
+      let _instαzeroone ← synthInstanceQ q(ZeroLEOneClass $α)
+      let _instαposmul ← synthInstanceQ q(PosMulStrictMono $α)
+      let _instαnontriv ← synthInstanceQ q(Nontrivial $α)
+      assertInstancesCommute
+      let pr : Q(∀ i, 0 < $f i) ← mkLambdaFVars #[i] pbody
+      pure $ .positive q(prod_pos fun i _ ↦ $pr i)
+    -- Try to show that the sum is nonnegative
+    catch _ => try
+      let pbody ← rbody.toNonneg
+      let pr : Q(∀ i, 0 ≤ $f i) ← mkLambdaFVars #[i] pbody
+      -- TODO: We must name the following, else `assertInstancesCommute` loops.
+      let _instαmon ← synthInstanceQ q(CommMonoidWithZero $α)
+      let _instαzeroone ← synthInstanceQ q(ZeroLEOneClass $α)
+      let _instαposmul ← synthInstanceQ q(PosMulMono $α)
+      assertInstancesCommute
+      pure $ .nonnegative q(prod_nonneg fun i _ ↦ $pr i)
+    catch _ =>
+      let pbody ← rbody.toNonzero
+      let pr : Q(∀ i, $f i ≠ 0) ← mkLambdaFVars #[i] pbody
+      -- TODO: We must name the following, else `assertInstancesCommute` loops.
+      let _instαmon ← synthInstanceQ q(CommMonoidWithZero $α)
+      let _instαnontriv ← synthInstanceQ q(Nontrivial $α)
+      let _instαnozerodiv ← synthInstanceQ q(NoZeroDivisors $α)
+      assertInstancesCommute
```

Comment:
```suggestion
    let rbody ← core zα pα body
    -- TODO: We must name the following, else `assertInstancesCommute` loops.
    let _instαmon ← synthInstanceQ q(CommMonoidWithZero $α)
    let _instαzeroone ← synthInstanceQ q(ZeroLEOneClass $α)
    let _instαposmul ← synthInstanceQ q(PosMulStrictMono $α)
    let _instαnontriv ← synthInstanceQ q(Nontrivial $α)
    let _instαnozerodiv ← synthInstanceQ q(NoZeroDivisors $α)
    assertInstancesCommute
    -- Try to show that the sum is positive
    try
      let .positive pbody := rbody | failure -- Fail if the body is not provably positive
      let pr : Q(∀ i, 0 < $f i) ← mkLambdaFVars #[i] pbody
      pure $ .positive q(prod_pos fun i _ ↦ $pr i)
    -- Try to show that the sum is nonnegative
    catch _ => try
      let pbody ← rbody.toNonneg
      let pr : Q(∀ i, 0 ≤ $f i) ← mkLambdaFVars #[i] pbody
      pure $ .nonnegative q(prod_nonneg fun i _ ↦ $pr i)
    catch _ =>
      let pbody ← rbody.toNonzero
      let pr : Q(∀ i, $f i ≠ 0) ← mkLambdaFVars #[i] pbody
      pure $ .nonzero q(prod_ne_zero fun i _ ↦ $pr i)
```
moving to earlier and deduplicating makes the tests still work locally: is this a reasonable change?

What I can see that might fail is if you some of those `synthInstanceQ` commands fails only in the `catch`, but then maybe there should be a test against that.

Note: I have no experience of working with `Qq`, so feel free to give profuse comments about why this is a bad idea for my own edification!

Reply:
Update: the suggested changes pass CI.

https://github.com/leanprover-community/mathlib4/actions/runs/8168589642/job/22330941550

Reply:
> What I can see that might fail is if some of those `synthInstanceQ` commands fails only in the `catch`, but then maybe there should be a test against that.

I'm not sure this is what you meant, but indeed your version has a problem if the body is strictly positive but multiplication isn't strictly monotone. I have now added a test that works with my version but not yours.

Reply:
Ok, this is exactly what I meant, thanks!